### PR TITLE
Add cell id per notebook format 4.5

### LINKED ISSUE #9645: cell id changes on save
The new version of `nbformat` has introduced cell ids.

I wanted to test the new functionality

## Reproduce

1. Opened a new jupyter notebook
2. Added a cell with content print("hello") and saved the notebook
3. Stored the cell id
4. Added a second cell with the content print("world") and saved the notebook

## Expected behavior

Compare the cell id of the first cell in step 4. with the one stored in step 3. they should be the same.

## Context

- Operating System and version: Windows 7
- Browser and version: Chrome Version 86.0.4240.183 
- JupyterLab version: 3.0.5

<details><summary>Troubleshoot Output</summary>
<pre>
The two ids are different.
</pre>
</details>

### LINKED ISSUE #9729: Support notebook format 4.5, adding cell IDs
We currently support notebook format 4.4. Notebook format 4.5 adds a new `id` field to cells. This issue is about supporting this newest version of the notebook format.

* [JEP](https://jupyter.org/enhancement-proposals/62-cell-id/cell-id.html)
* [nbformat PR](https://github.com/jupyter/nbformat/pull/189)
* [nbformat documentation](https://nbformat.readthedocs.io/en/latest/format_description.html#cell-ids)

### Related issues:

* Some conversation in https://github.com/jupyterlab/jupyterlab/issues/9064 is related to implementing this.
* https://github.com/jupyterlab/jupyterlab/issues/9645 details unexpected behavior with format 4.5 files
* https://github.com/jupyter/nbformat/issues/209 also discussed jlab support

### PR DESCRIPTION
**Pull request by jayqi**
## References

Closes #9729. Fixes #9645.  

## Code changes

- Adds `id` field to `RawCellModel`,  `CodeCellModel`, and `MarkdownCellModel`'s interface and `toJSON` methods to support the cell IDs of notebook format 4.5

---  

@jasongrout I've been poking around at adding cell IDs and could use some help. I'm not very familiar with the Jupyter Lab codebase, and I'm also not much of a Javascript/Typescript programmer. 

One of the things I've encountered is that the cell models already had an internal `id` field that's a UUID, so there's some potential collision there that I don't really know the best way to handle. Maybe the new format's cell IDs should replace the UUIDs? 

Currently, I'm successfully writing stable IDs to notebook files on save (they're not being regenerated), but it seems to be the UUIDs rather than the JEP ids. Somewhere it seems like they're being dropped on deserialization, or maybe overwritten? Not sure what I'm missing.

### CODE REVIEW

**Comment by echarles** [comment]
Thx a lot @jayqi for this. I have not tested your changes, but don't see any changes in the nbformat package. I don't think this would be the cause of the serialization issue you mentioned (uuid instead of id), but it is worth to mention. In any case, nbformat package should have the id attributed (if it has it not yet).

File: packages/cells/src/model.ts
Comment (jasongrout) [inline packages/cells/src/model.ts:428]: Can we do this in the super class's method, the one we are calling with `super.toJSON()`?

File: packages/nbformat/src/index.ts
Comment (jasongrout) [inline packages/nbformat/src/index.ts:224]: Can we add this field to the `IBaseCell`? Since it is the same for every type of cell, we don't need to specialize this field in each subinterface.

**Review by jasongrout — COMMENTED** [review]
Thanks! I think this is a good start!

We'll need to make sure that we are setting the major/minor version of the notebook to 4.5 as well when we save if there are ids, and not if there are no ids (or perhaps we generate ids if there aren't any, and always set the format to 4.5?). See https://github.com/jupyterlab/jupyterlab/blob/bc55a336f18e964995a5cfc7f3a9c2a71e1388ca/packages/nbformat/src/index.ts#L22

**Comment by jasongrout** [comment]
> One of the things I've encountered is that the cell models already had an internal `id` field that's a UUID, so there's some potential collision there that I don't really know the best way to handle. Maybe the new format's cell IDs should replace the UUIDs?

The cell ids seem to be set at https://github.com/jupyterlab/jupyterlab/blob/bc55a336f18e964995a5cfc7f3a9c2a71e1388ca/packages/cells/src/model.ts#L167

It could come from the `cell` as done with several other things below it. Is anyone calling this function and passing in an id? For backwards compatibility, what do you think of making that line something like this?

```typescript
this.id = options.id || cell.id || UUID.uuid4();
```
(and of course moving it down below where `cell` is defined, or defining `cell` just above this line)

File: packages/cells/src/model.ts
Comment (jayqi) [reply packages/cells/src/model.ts:428]: The reason I had done it this way was because the notebook 4.5 specification only has the `id` field on raw cells, code cells, and markdown cells. It's not part of the specification for urecognized cells. 

https://github.com/jupyter/nbformat/blob/master/nbformat/v4/nbformat.v4.5.schema.json

File: packages/nbformat/src/index.ts
Comment (jayqi) [reply packages/nbformat/src/index.ts:224]: See previous comment about `id` not being on the specification for unrecognized cells.

**Comment by jasongrout** [comment]
Just curious - does nbformat report that notebooks it is adding the ids to are format 4.5? After these changes, does JLab recognize this and tell the user that if they save, they'll be converting the notebook to format 4.5?

Ideally, if a user is opening a notebook without ids (4.4 format or earlier), jlab will tell people that it is automatically converting the notebook to format 4.5, and save the notebook with ids when it is saved.

**Comment by jasongrout** [comment]
On the other hand, as far as migration goes during this transition time, https://github.com/jupyter/nbformat/pull/189#issuecomment-736917847 is a sound strategy.

**Comment by jasongrout** [comment]
@jayqi - thanks again for working on this. I'd like to push forward on this together, if you'd like. Would you like to pair program on this issue, or if you've run out of time investigating this, do you mind if I push some commits to this PR?

**Comment by jasongrout** [comment]
I think these scenarios need to be tested:

- nbformat 5.0.8 installed, create new notebook: should create a notebook with format 4.4 and no cell ids
- nbformat 5.0.8 installed, open a 4.4 notebook: should preserve the notebook as format 4.4 and no cell ids
- nbformat 5.0.8 installed, open a 4.5 notebook: probably should complain that the notebook version is not supported?

- nbformat 5.1.2 installed, create a new notebook: should create a notebook with format 4.5 and cell ids
- nbformat 5.1.2 installed, open a 4.4 notebook: likely nbformat will force an upgrade to format 4.5 and auto-generate cell ids, so in that case warn the user about the upgrade if they save. If save happens, the notebook is saved as a 4.5 notebook preserving the nbformat-provided cell ids.
- nbformat 5.1.2 installed, open a 4.5 notebook: should preserve the notebook as format 4.5 and preserve any existing cell ids.

I think the creation cases can be handled like the opening cases, since creating a notebook is essentially asking the server to create it and then us requesting what the server created. In other words, even for creating a new notebook, the notebook model should reflect whether it is a 4.4 notebook for a 4.5 notebook.

I think in order to have the above behaviors, it might be sufficient for us to just write the cell ids if the minor format in the notebook model is 5, and not if the minor format is 4.

**Comment by jayqi** [comment]
Hi @jasongrout, I was just coincidentally coming back to this PR just now. I tried your suggestion of  

```
this.id = options.id || cell.id || UUID.uuid4();
```

but it appears to fail the static type-checking? 

```
../cells/src/model.ts:167:5 - error TS2322: Type 'string | number | true | PartialJSONObject | PartialJSONArray' is not assignable to type 'string'.
  Type 'number' is not assignable to type 'string'.

167     this.id = options.id || options.cell.id || UUID.uuid4();
        ~~~~~~~

../cells/src/model.ts:167:29 - error TS2532: Object is possibly 'undefined'.

167     this.id = options.id || options.cell.id || UUID.uuid4();
                                ~~~~~~~~~~~~
```

I do have some time later tonight (I'm in U.S. PDT) to work on this, or, coincidentally, tomorrow is a hack day at my work so I could also potentially work on this during work hours. Pairing would be helpful because I'm a bit in over my head here. I'm also open to having you push commits directly—I'm more interested in the problem being resolved than being the one whose name is on the commits.

**Comment by jasongrout** [comment]
> I do have some time later tonight (I'm in U.S. PDT) to work on this, or, coincidentally, tomorrow is a hack day at my work so I could also potentially work on this during work hours. Pairing would be helpful because I'm a bit in over my head here. I'm also open to having you push commits directly—I'm more interested in the problem being resolved than being the one whose name is on the commits.

Awesome. I'm about ready to quit for the day, and tomorrow is a holiday for my work, so I'll be getting back to this Monday. Do you want to pair on Monday?

I have some work-in-progress commits over at my branch at https://github.com/jasongrout/jupyterlab/tree/cell-id that build on your commits, as experiments in pushing things forward a bit more. Feel free to take a look, and I'll check the comments here before looking at things again on Monday.

Thanks again also for opening this PR and starting the work on this.

**Comment by jayqi** [comment]
I pulled your changes and tried it out, and here are my observations:

- When creating a new notebook, I'm getting a UUID written as the cell ID. 
- When I save, the cell ID stays the same UUID. 
- When I stop the kernel, reopen the notebook, and save, the cell ID stays the same UUID.
- When I shut down Jupyter Lab and restart it, and then save the notebook, a new UUID is generated. 
- When I edit the notebook manually and paste in a format 4.5 style cell ID (like `"opposite-spectrum"`), and then I open the notebook in Jupyter Lab and save, the cell ID is replaced by a UUID.
- When I edit the notebook manually and set `nbformat_minor` to `4`, and then I open the notebook in Jupyter Lab and save, the cell ID is replaced by a new UUID (even if it was a UUID already) and the `nbformat_minor` value is replaced as `5`. 

I think Monday should work for me. I'd prefer evening after work (since this isn't my job) but if it needs to be work hours, my afternoon schedule is currently pretty open and I can take some time for this.

**Comment by jasongrout** [comment]
> When I edit the notebook manually and paste in a format 4.5 style cell ID (like `"opposite-spectrum"`), and then I open the notebook in Jupyter Lab and save, the cell ID is replaced by a UUID.

This is happening at https://github.com/jupyterlab/jupyterlab/blob/c0316179d8dda04e11335074842942127e76b72a/packages/notebook/src/model.ts#L522 (and the similar lines in createMarkdownCell and createRawCell just below this)

**Comment by jasongrout** [comment]
> When I edit the notebook manually and paste in a format 4.5 style cell ID (like `"opposite-spectrum"`), and then I open the notebook in Jupyter Lab and save, the cell ID is replaced by a UUID.

I think I fixed this with the latest commit. Can you test to see if this still happens?

File: packages/cells/src/model.ts
Comment (jasongrout) [inline packages/cells/src/model.ts:167]: Test:
- [ ] If a cell is created without an option id, but does have a cell id, that is the id that is used for the cell

File: packages/cells/src/model.ts
Comment (jasongrout) [inline packages/cells/src/model.ts:428]: Test:
- [ ] a markdown cell preserves its ids in its toJSON output
- [ ] a raw cell preserves its ids in its toJSON output
- [ ] a code cell preserves its ids in its toJSON output

File: packages/notebook/src/model.ts
Comment (jasongrout) [inline packages/notebook/src/model.ts:198]: Tests:
- [ ] if format is 4.4, cell ids are not saved
- [ ] if format is 4.5, cell ids are saved

File: packages/notebook/src/model.ts
Comment (jasongrout) [inline packages/notebook/src/model.ts:229]: Tests:
- [ ] if notebook format is 4.5, cell ids in a notebook are preserved from the input JSON to the notebook model for code, markdown, and raw cells.

**Review by jasongrout — COMMENTED** [review]
I added some comments about tests that should be written.

**Comment by jasongrout** [comment]
I think this may be ready for review now. @jayqi - what do you think of my latest commits and tests?

**Comment by jasongrout** [comment]
Because the issue fixed here is causing a lot of pain for people, I'd suggest either backporting this to 3.0.x if it will be a while before 3.1 is released, or releasing 3.1 very soon to get this fix out there.

**Review by blink1073 — APPROVED** [review]

**Comment by blink1073** [comment]
> I'd suggest either backporting this to 3.0.x

👍🏼 

Happy to merge and backport once you are ready @jayqi

**Comment by jayqi** [comment]
@jasongrout I just pulled and ran the changes locally, and one remaining thing I'm noticing is that new cells are being generated with UUIDs for the cell ID rather than the random human-readable strings that I get with v3.0.12. I guess from reading the [relevant section of the JEP](https://github.com/jupyter/enhancement-proposals/blob/master/62-cell-id/cell-id.md#recommended-application--usage-of-id-field), UUIDs are acceptable though the human-readable strings are preferred? It may be valuable to match the recommended implementation in the JEP / what other applications in the Jupyter ecosystem are doing, but I will defer to you all on that. 

Otherwise, this all looks reasonable to me (with the caveat that this is fairly over my head). Thanks for picking it up @jasongrout and getting it done.

**Comment by blink1073** [comment]
> new cells are being generated with UUIDs for the cell ID rather than the random human-readable strings 

This was a recent change in `nbformat` itself: https://github.com/jupyter/nbformat/pull/217

**Comment by blink1073** [comment]
@meeseeksdev please backport to 3.0.x

**Pull request merged**
